### PR TITLE
.*: add more dm-ctl info in ctl operations

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
Current dm-ctl operation are async. Users need to query-status and get results after they do the operations which is not easy.

### What is changed and how it works?
dm-master will run `query-status` and check whether the result is correct and return the simplified result to users.

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `dm/dm-ansible`
 - Need to be included in the release note

### CODE REVIEW

**Comment by lichunzhu** [comment]
@WangXiangUSTC @csuzhangxc PTAL

File: dm/master/server.go [hunk 1]
```diff
@@ -1334,37 +1341,232 @@ func (s *Server) generateSubTask(ctx context.Context, task string) (*config.Task
...
 
-// taskConfigArgsExtractor extracts SubTaskConfig from args and returns its relevant
-// grpc client, worker id (host:port), subtask config in toml, task name and error
-func (s *Server) taskConfigArgsExtractor(cfg *config.SubTaskConfig) (*scheduler.Worker, string, string, error) {
-	handleErr := func(err error) error {
-		log.L().Error("response", zap.Error(err))
-		return err
+func extractWorkerError(result *pb.ProcessResult) error {
+	if result != nil && len(result.Errors) > 0 {
+		return terror.ErrMasterOperRespNotSuccess.Generate(result.Errors[0].String())
```
Comment (WangXiangUSTC) [inline dm/master/server.go]: how about adding all error's information in `result.Errors`?

File: dm/master/server.go [hunk 2]
```diff
@@ -1334,37 +1341,232 @@ func (s *Server) generateSubTask(ctx context.Context, task string) (*config.Task
...
+		},
 	}
 
-	worker := s.scheduler.GetWorkerBySource(cfg.SourceID)
-	if worker == nil {
-		return nil, "", "", handleErr(terror.ErrMasterTaskConfigExtractor.Generatef("%s relevant worker-client not found", cfg.SourceID))
+	for num := 0; num < maxRetryNum; num++ {
+		// check whether source relative worker has been removed by scheduler
+		switch masterReq.(type) {
+		case *pb.OperateSourceRequest:
```
Comment (WangXiangUSTC) [inline dm/master/server.go]: seems that no need to use `switch`

File: dm/master/server.go (hunk 1 above)
Comment (lichunzhu) [reply dm/master/server.go]: revised in https://github.com/pingcap/dm/pull/490/commits/6d43a1ddce1f583c1a734b38e787d65cc9ad6261

File: dm/master/server.go (hunk 2 above)
Comment (lichunzhu) [reply dm/master/server.go]: revised in https://github.com/pingcap/dm/pull/490/commits/6d43a1ddce1f583c1a734b38e787d65cc9ad6261

File: dm/worker/server.go [hunk 3]
```diff
@@ -210,20 +212,30 @@ func (s *Server) setWorker(worker *Worker, needLock bool) {
...
 }
 
-func (s *Server) setWorkerErr(workerErr error, needLock bool) {
+func (s *Server) setSourceStatus(source string, err error, needLock bool) {
 	if needLock {
 		s.Lock()
 		defer s.Unlock()
 	}
-	s.workerErr = workerErr
+	s.sourceStatus = pb.SourceStatus{
```
Comment (csuzhangxc) [inline dm/worker/server.go]: Do we never set `RelayStatus` for `s.sourceStatus`, right?

File: dm/worker/server.go [hunk 4]
```diff
@@ -435,8 +447,7 @@ func (s *Server) QueryStatus(ctx context.Context, req *pb.QueryStatusRequest) (*
 	}
 
 	resp.SubTaskStatus = w.QueryStatus(req.Name)
-	resp.RelayStatus = relayStatus
-	resp.Source = w.cfg.SourceID
+	sourceStatus.RelayStatus = relayStatus
```
Comment (csuzhangxc) [inline dm/worker/server.go]: How about setting `RelayStatus` to `nil` to make the response of `query-status` clearer when `enable-relay: false`? cc @WangXiangUSTC .

File: dm/worker/server.go [hunk 5]
```diff
@@ -446,24 +457,30 @@ func (s *Server) QueryStatus(ctx context.Context, req *pb.QueryStatusRequest) (*
 // QueryError implements WorkerServer.QueryError
 func (s *Server) QueryError(ctx context.Context, req *pb.QueryErrorRequest) (*pb.QueryErrorResponse, error) {
 	log.L().Info("", zap.String("request", "QueryError"), zap.Stringer("payload", req))
+	sourceStatus := s.getSourceStatus(true)
+	sourceError := pb.SourceError{
+		Source: sourceStatus.Source,
+		Worker: s.cfg.Name,
+	}
+	if sourceStatus.Result != nil && len(sourceStatus.Result.Errors) > 0 {
+		sourceError.SourceError = sourceStatus.Result.Errors[0].String()
```
Comment (csuzhangxc) [inline dm/worker/server.go]: Can we `join` all `Result.Errors` here? or it will have at most one error?

File: dm/worker/server.go (hunk 3 above)
Comment (lichunzhu) [reply dm/worker/server.go]: Yes

Comment (csuzhangxc) [reply dm/worker/server.go]: Add a comment to clarify it?

File: dm/master/server.go [hunk 6]
```diff
@@ -305,6 +307,7 @@ func subtaskCfgPointersToInstances(stCfgPointers ...*config.SubTaskConfig) []con
 func (s *Server) StartTask(ctx context.Context, req *pb.StartTaskRequest) (*pb.StartTaskResponse, error) {
 	log.L().Info("", zap.Stringer("payload", req), zap.String("request", "StartTask"))
 
+	resp := &pb.StartTaskResponse{}
```
Comment (csuzhangxc) [inline dm/master/server.go]: You can declare this variable after `isLeaderAndNeedForward` check.

File: dm/worker/server.go (hunk 5 above)
Comment (lichunzhu) [reply dm/worker/server.go]: addressed in https://github.com/pingcap/dm/pull/490/commits/b9f4793fa1b4d18afa02c9fd9987ce42991dad44

File: dm/worker/server.go (hunk 3 above)
Comment (lichunzhu) [reply dm/worker/server.go]: addressed in https://github.com/pingcap/dm/pull/490/commits/b9f4793fa1b4d18afa02c9fd9987ce42991dad44

File: dm/worker/server.go (hunk 4 above)
Comment (WangXiangUSTC) [reply dm/worker/server.go]: I think it's ok

**Pull request merged**